File: haskell/lib/Unification.hs
type Id = String

data Term = Var Id | Term Id [Term]

type Substitution = [(Id, Term)]

occurs :: Id -> Term -> Bool
occurs x (Var y) = x == y
occurs x (Term _ ts) = any (occurs x) ts

subst :: Term -> Id -> Term -> Term
subst s x (Var y) = if x == y then s else Var y
subst s x (Term f ts) = Term f (map (subst s x) ts)

apply :: Substitution -> Term -> Term
apply s t = foldr (\(x, u) -> subst u x) t s

unifyOne :: Term -> Term -> Substitution
unifyOne (Var x) (Var y)
  | x == y = []
  | otherwise = [(x, Var y)]
unifyOne (Term f sc) (Term g tc)
  | f == g && length sc == length tc = unify $ zip sc tc
  | otherwise = error "not unifiable: head symbol conflict"
unifyOne (Var x) t@(Term _ _)
  | occurs x t = error "not unifiable: circularity"
  | otherwise = [(x, t)]
unifyOne t@(Term _ _) (Var x)
  | occurs x t = error "not unifiable: circularity"
  | otherwise = [(x, t)]

unify :: [(Term, Term)] -> Substitution
unify [] = []
unify ((x, y):ts) = let t2 = unify ts
                        t1 = unifyOne (apply t2 x) (apply t2 y)
                    in t1 ++ t2

File: python/unification.py
from typing import Dict, List, Tuple

Id = str

class Term:
    def __init__(self, id: Id, args: List['Term'] = None):
        self.id = id
        self.args = args or []

    def isVar(self) -> bool:
        return not self.args

    def __repr__(self):
        if self.isVar():
            return self.id
        else:
            return f"{self.id}({', '.join(repr(arg) for arg in self.args)})"

Substitution = Dict[Id, Term]

def occurs(x: Id, t: Term) -> bool:
    if t.isVar():
        return x == t.id
    else:
        return any(occurs(x, arg) for arg in t.args)

def subst(s: Term, x: Id, t: Term) -> Term:
    if t.isVar():
        return s if x == t.id else t
    else:
        return Term(t.id, [subst(s, x, arg) for arg in t.args])

def apply(s: Substitution, t: Term) -> Term:
    if t.isVar():
        return s.get(t.id, t)
    else:
        return Term(t.id, [apply(s, arg) for arg in t.args])

def unifyOne(s: Term, t: Term) -> Substitution:
    if s.isVar():
        if t.isVar():
            return {s.id: t} if s.id != t.id else {}
        elif occurs(s.id, t):
            raise ValueError("Not unifiable: circularity")
        else:
            return {s.id: t}
    elif t.isVar():
        if occurs(t.id, s):
            raise ValueError("Not unifiable: circularity")
        else:
            return {t.id: s}
    elif s.id != t.id or len(s.args) != len(t.args):
        raise ValueError("Not unifiable: head symbol conflict")
    else:
        sub = {}
        for s_arg, t_arg in zip(s.args, t.args):
            sub.update(unifyOne(apply(sub, s_arg), apply(sub, t_arg)))
        return sub

def unify(pairs: List[Tuple[Term, Term]]) -> Substitution:
    sub = {}
    for s, t in pairs:
        sub.update(unifyOne(apply(sub, s), apply(sub, t)))
    return sub